drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1454 @@
+ * for now in case we find another BSOD.
+ */
+
+#define ExAllocatePoolWithTag(pool, size, tag) \
+({ \
+	void *ret; \
+	if (tag == TAG_NETIO) { \
+		ret = ExAllocatePoolWithTag(pool, size+4, tag); \
+		if (ret != NULL) \
+			*(int*)(((char*)ret)+size) = 0xabcdef; \
+	} else { \
+		ret = ExAllocatePoolWithTag(pool, size, tag); \
+	} \
+        NETIO_DbgPrint(MIN_TRACE, ("ExAllocatePoolWithTag returning %p size is %d tag is %x\n", ret, size, tag)) \
+	ret; \
+})
+
+#define ExFreePoolWithTag(p, tag) \
+	NETIO_DbgPrint(MIN_TRACE, ("ExFreePoolWithTag %p %x\n", p, tag)) \
+	ExFreePoolWithTag(p, tag)
+
+#endif
+
+typedef struct _WSK_SOCKET_INTERNAL
+{
+    WSK_SOCKET s;
+
+    ADDRESS_FAMILY family;      /* AF_INET or AF_INET6 */
+    USHORT type;                /* SOCK_DGRAM, SOCK_STREAM, ... */
+    ULONG proto;                /* IPPROTO_UDP, IPPROTO_TCP */
+    ULONG flags;                /* WSK_FLAG_LISTEN_SOCKET, ... */
+    PVOID user_context;         /* parameter for callbacks, opaque */
+    UNICODE_STRING TdiName;     /* \\Devices\\Tcp, \\Devices\\Udp */
+
+    struct sockaddr LocalAddress;
+    PFILE_OBJECT LocalAddressFile;
+    HANDLE LocalAddressHandle;
+
+    struct sockaddr RemoteAddress;      /* Remeber latest remote connection */
+    PFILE_OBJECT RemoteAddressFile;
+    HANDLE RemoteAddressHandle;
+
+    /* Those exist for connection oriented (TCP/IP) sockets only */
+    PFILE_OBJECT ConnectionFile;        /* Returned by TdiOpenConnectionEndpointFile() */
+    HANDLE ConnectionHandle;            /* Returned by TdiOpenConnectionEndpointFile() */
+    BOOLEAN ConnectionFileAssociated;
+
+    /* Incoming connection callback function: */
+    const WSK_CLIENT_LISTEN_DISPATCH *ListenDispatch;
+    UINT CallbackMask;
+
+    UINT Flags;                          /* SO_REUSEADDR, ... see ws2def.h */
+    UINT RefCount;                       /* See SocketGet/SocketPut TODO: this should be atomic */
+
+    PIRP ListenIrp;	           /* must be cancelled on close */
+    BOOLEAN ListenCancelled;
+    HANDLE ListenThreadHandle;     /* needed to restart listening */
+    PKTHREAD ListenThread;
+    KEVENT StartListenEvent;
+    BOOLEAN ListenThreadShouldRun;
+
+    /* AcceptSocket's keep a reference on their listen sockets so
+     * that the Address file will be closed only if there are no
+     * open connections any more.
+     */
+    struct _WSK_SOCKET_INTERNAL *ListenSocket;
+
+    /* SocketPut() only works at PASSIVE_LEVEL. If > PASSIVE_LEVEL,
+     * it will put the socket on this list and wake the putsockets
+     * thread.
+     */
+    struct _WSK_SOCKET_INTERNAL *NextSocketToPut;
+    int NumSocketPuts;
+} WSK_SOCKET_INTERNAL, *PWSK_SOCKET_INTERNAL;
+
+struct NetioContext
+{
+    PIRP UserIrp;
+    PWSK_SOCKET_INTERNAL socket;
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo;
+    PTDI_CONNECTION_INFORMATION PeerAddrRet;
+};
+
+void
+SocketGet(PWSK_SOCKET_INTERNAL s)
+{
+    FUNCTION_TRACE;
+
+    s->RefCount++;
```

Comment:
```suggestion
    InterlockedIncrement(&s->RefCount);
```
It may be also worth guarding against wrapping after the incrementation...

Reply:
Hi, in theory it is necessary everywhere a InterlockedIncrement is called. Not sure what I should do in that case, however, except maybe print a warning and wait for the machine to BSOD later (double free in SocketPut). Hmm..